### PR TITLE
Timestep bias for fine-tuning SDXL

### PR DESCRIPTION
**Pull request by bghira**
# What does this PR do?

This change introduces parameters that allow the user to direct the model toward a certain subset of timesteps more frequently.

Empirically, we've observed more rapid convergence with this change and an improvement in fine details when using the `later` timestep bias strategy and a weight multiplier of 5.

Similarly, we've observed more rapid changes to the composition of the model by using the `early` strategy.

This has been monumentally useful in the quest to tune a terminal SNR variant.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by bghira** [comment]
prompt: `a thought-provoking portrait of a masked dancer, shot with a Canon EOS 5D Mark IV on f/2`

Before, how things looked for about 7200 steps of training. Difficult to move past some issues.

image

After using `--timestep_bias_strategy=later --timestep_bias_portion=.25 --timestep_bias_multiplier=5` for 400 steps:

image

**Comment by bghira** [comment]
another stubborn case that just seemed to be degrading rather than improving.

prompt: `an aerial view of a lush green forest, photographed with a DJI Mavic Pro drone at 200 feet`

image

after 400 steps:

image

**Comment by bghira** [comment]
image

loss chart showing the change during training toward the end.

**Comment by patrickvonplaten** [comment]
Very interesting! Does this originate from a paper / some research github?

**Comment by bghira** [comment]
this is my team's research outcome. i am just glad it works, hahah, it has greatly reduced the financial cost of fine-tuning

**Comment by bghira** [comment]
unexpected gains!

the final 25% of timesteps seem responsible for, eg. horizon continuity:

image

after training:

image

**Comment by bghira** [comment]
preliminary testing of this change in the LoRA trainer also leads to a major reduction in time taken to train the network, with lower risk of "frying" especially when using the trailing 100 timesteps rather than the trailing 200 timesteps.

**Comment by patrickvonplaten** [comment]
Very interesting! @sayakpaul @williamberman @patil-suraj wdyt?

**Comment by bghira** [comment]
more testing with this strategy for LoRAs.

we tested a hard cut-off of steps the LoRA trains for, such that the network only learnt the final 50, 100, 200 steps, and then the first 200 steps.

the hard cut-off is heavily data-dependent, as even limiting it to the final 200 steps performs poorly compared to 100 steps.

however, all of the resulting LoRAs were an improvement over the base model. it's only with respect to each other that performance was up or down.

when training LoRAs using bias instead of a hard cut-off, this issue isn't as pronounced, presumably because the network "learns" a few other useful timesteps.

**Comment by patrickvonplaten** [comment]
From a first look this makes a lot of sense to me. Some questions:
- Did you notice that when the model is not **at all** trained on say the later timesteps (500-1000) that the image quality for just these timesteps is degraded?
- In some sense this method could also allow to train/fine-tune a refiner-like checkpoint quite easily (e.g. only train on timesteps 0-200 - did you try that? E.g. training a checkpoint just for img2img?

**Comment by bghira** [comment]
it has indeed occurred to me that this approach could majorly improve img2img capabilities. especially then experimenting with block weight merge to constrain the changed weights to the out blocks instead of the inputs. a lot of the changes we make don't need to happen to the whole model. i've even considered freezing portions of the u-net to try and adapt the noise schedule more cleanly, but it's not something i really know how to do effectively.

if i had more compute available, i would certainly be running those experiments, but i am limited to two concurrent long training sessions on my two 80G GPUs and then short burn tests on an occasional multi-GPU system. it's quite difficult to actually rent time on a multi-A100 80G system due to lack of availability, and when I use A6000s at 48G each, I can't realistically make use of EMA weights at 1024px. it doesn't feel like the wisest use of resources.


i generated a scatter plot of the timesteps being trained over 1000 steps and then merged that along the X-axis and saw a fairly strong distribution around the middle of the schedule, but the edges are hardly touched.

we then made a timer that kept track of how often the terminal timestep itself was even reached during training, and it was once every 350 samples. that's not often enough at all to really learn this new schedule.

also itching to bias the training on earlier timesteps with some different style datasets and see how that impacts the style transfer and so on.

File: examples/text_to_image/train_text_to_image_sdxl.py [hunk 1]
```diff
@@ -325,6 +325,55 @@ def parse_args(input_args=None):
     parser.add_argument(
         "--lr_warmup_steps", type=int, default=500, help="Number of steps for the warmup in the lr scheduler."
     )
+    parser.add_argument(
+        "--timestep_bias_strategy",
+        type=str,
+        default="none",
+        choices=["earlier", "later", "none"],
+        help=(
+            "The timestep bias strategy, which may help direct the model toward learning low or frequency details."
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_sdxl.py:334]: "low or frequency details" sounds a bit off to me.

File: examples/text_to_image/train_text_to_image_sdxl.py [hunk 2]
```diff
@@ -325,6 +325,55 @@ def parse_args(input_args=None):
...
+            " Choices: ['earlier', 'later', 'none']."
+            " The default is 'none', which means no bias is applied, and training proceeds normally."
+            " The value of 'later' will prefer to generate samples for later timesteps."
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_sdxl.py:337]: I don't understand what "generate samples for later timesteps" means.

File: examples/text_to_image/train_text_to_image_sdxl.py [hunk 3]
```diff
@@ -325,6 +325,55 @@ def parse_args(input_args=None):
...
+            "The multiplier for the bias. Defaults to 1.0, which means no bias is applied."
+            " A value of 2.0 will double the weight of the bias, and a value of 0.5 will halve it."
+        ),
+    )
+    parser.add_argument(
+        "--timestep_bias_begin",
+        type=int,
+        default=0,
+        help=(
+            "When using `--timestep_bias_strategy=range`, the beginning (inclusive) timestep to bias."
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_sdxl.py:354]: But the `timestep_bias_strategy` doesn't have this choice.

File: examples/text_to_image/train_text_to_image_sdxl.py [hunk 4]
```diff
@@ -935,11 +1025,12 @@ def collate_fn(examples):
...
-                # Sample a random timestep for each image
-                timesteps = torch.randint(
-                    0, noise_scheduler.config.num_train_timesteps, (bsz,), device=model_input.device
+                # Sample a random timestep for each image, potentially biased by the timestep weights.
+                # Biasing the timestep weights allows us to spend less time training irrelevant timesteps.
+                weights = generate_timestep_weights(args, noise_scheduler.config.num_train_timesteps).to(
+                    model_input.device
                 )
-                timesteps = timesteps.long()
+                timesteps = torch.multinomial(weights, bsz, replacement=True).long()
```
Comment (sayakpaul) [inline examples/text_to_image/train_text_to_image_sdxl.py:1033]: Could we maybe condition this block so as to not get rid of `torch.randint()` in the interest of readability? Our scripts are referred by the community to develop understanding too. I believe the timestep sampling part during training (albeit simplistic) is a crucial part there. So, I'd prefer to slightly condition this block.

**Comment by sayakpaul** [comment]
Thanks so much for sharing your insights here! Always appreciated. 

@kashif and I experimented with the idea of timestep annealing earlier this year but couldn't find significant improvements in the aesthetic quality of the results. But the loss landscape improve quite a bit, though. Our idea was similar i.e., to enforce a way for the model to progress in a certain timestep range based on where it's in the current training process.

File: examples/text_to_image/train_text_to_image_sdxl.py (hunk 2 above)
Comment (bghira) [reply examples/text_to_image/train_text_to_image_sdxl.py]: i have rewritten this now.

File: examples/text_to_image/train_text_to_image_sdxl.py (hunk 1 above)
Comment (bghira) [reply examples/text_to_image/train_text_to_image_sdxl.py]: ah yes, this should have said "or high"

File: examples/text_to_image/train_text_to_image_sdxl.py (hunk 3 above)
Comment (bghira) [reply examples/text_to_image/train_text_to_image_sdxl.py:354]: oops. it actually does internally, but after adding that, i forgot to update the documentation. thanks for the catch!

**Comment by bghira** [comment]
@sayakpaul let me know if those changes meet your satisfaction, if not for the wording changes then please provide some alternative wording that would work for you - and i can port these changes to the other PRs

i think maybe for most training this won't have much of an impact, but i'm probably going to do extensive testing with it in SD 2.0 and SD 2.1 at some point once the GPUs are freed up from SDXL.

**Comment by sayakpaul** [comment]
Once the code quality issues are addressed I am game for merging this!

**Comment by bghira** [comment]
```
Run black --check examples tests src utils scripts
13
All done! ✨ 🍰 ✨
14
646 files would be left unchanged.
15
/opt/actions-runner/_work/_temp/4a988597-e0f7-4430-b7aa-4aff4f890a69.sh: line 3: doc-builder: command not found
16
Error: Process completed with exit code 127.
```

not sure that's related.

**Comment by patrickvonplaten** [comment]
@sayakpaul can you check?

**Comment by sayakpaul** [comment]
@bghira I will merge the PR today. But I want to also welcome you to write a short blog post on this topic and discuss the results. We're happy to publish it from hf.co/blog. We can then definitely mention it from our training docs.

**Comment by bghira** [comment]
@sayakpaul ok i will include that with some checkpoints and make those available to you this weekend
